Prevent import cycles via assumptions.ask

toolforger opened the pull request:
ask.py would collect known facts during import, calling back into SymPy.
Replaced the statements with cached functions that would do it, so the
collection passes will happen only after SymPy is fully loaded.

The exec-based consistency test would not work for a function-based
facts collection; for some reason, it would not grok the import of
Or (and most likely other symbols).
I suspect that the global namespace is lost somewhere along the way
from exec String to cache to function, but I could not find that out;
instead, I resorted to a textual compare of ask_generated.py with what
compute_known_facts() returns.
If the consistency test is just to prevent an outdated ask_generated.py,
then that works, too. If the original consistency test prevented other
mishaps not tested for elsewhere, then that needs fixing.

toolforger commented:
Performance:
```bin/test``` before 4795.24 s, after 4797.57 s
```bin/test assumptions``` before 55.23 s, after 56.66 s
next perf test: ```bin/test assumptions``` before 60.65 s, after 58.03 s

toolforger commented:
@mrocklin @asmeurer @rlamy ping?

jcrist commented:
I'm surprised that there's such little performance difference in this. I'd expect more of a hit for repeated function calls. That said, this still seems like an ugly hack, and there has to be a better way. @asmeurer or @debugger22 might have better thoughts.

toolforger commented:
Actually the performance difference is below the noise threshold.
Possible explanations:
1) The function call happens just during setup of a run of facts inference. The impact may be too low to be measurable.
2) There may be an override of `__getattr__` etc. up the inheritance chain. If so (I haven't checked), then any attribute access is a function call anyway. In fact the function call could be minimally faster than the attribute access because the Python interpreter would not need to check that many override hooks.

Which aspect of this do you consider hackish?
The string comparison, or the attribute-to-function-call transformation?

jcrist commented:
> Which aspect of this do you consider hackish?
The string comparison, or the attribute-to-function-call transformation?

The string comparison actually seems better than what we were doing before - the function call is what I'm not so sure about. But barring a major refactor of the assumptions system (might happen soon anyway, with GSoC and all that), I can't think of a better way to remove the import cycles.

Two things that would make me feel better about this change:

1. Rename the functions to better imply that they're functions. `known_facts_dict` is no longer a `dict` - perhaps `get_known_facts_dict`.

2. Call the functions just once in each body (i.e. `known_facts_dict = get_known_facts_dict()`). This may not seem necessary from the benchmark, but we're not stressing this assumption system very much right now (it's not heavily used yet). It still may not be critical even then, but avoiding repeated function calls would make me feel better:)

toolforger commented:
> The string comparison actually seems better than what we were doing before

Now that we're talking about it, I guess ideally we want to compare parses.
OTOH what we have here is good enough I suppose :-)

 > - the function call is what I'm not so sure about. But barring a 
major refactor of the assumptions system (might happen soon anyway, with 
GSoC and all that), I can't think of a better way to remove the import 
cycles.

Class variables is that they get initialized during class definition. If 
the initialization calls into SymPy, there's a 50% likelihood that that 
call is hitting some not-yet-initialized portion of SymPy.

That's a recurring theme. Even if the new assumptions system removes all 
the existing cycles, it will likely come with its own. Or, more 
deviously, with almost-completed cycles that won't become visible until 
some other part of SymPy is changed and accidentally closes the cycle.

Lesson to take home: Access class variables only through functions, that 
way the class variables can be initialized lazily.
We amplify this problem to real-world proportions through Singleton. 
Singleton needs to create the instance during class creation, so the 
problem extends to member variables.

> 1. Rename the functions to better imply that they're functions. `known_facts_dict` is no longer a `dict` - perhaps `get_known_facts_dict`.

Okay, I can do that.

> 2. Call the functions just once in each body (i.e. `known_facts_dict = get_known_facts_dict()`). This may not seem necessary from the benchmark, but we're not stressing this assumption system very much right now (it's not heavily used yet). It still may not be critical even then, but avoiding repeated function calls would make me feel better:)

Oh, there are repeated calls? Usually I avoid them. I'll take a look.

toolforger commented:
The rename is done, repeated call checking is going next.

toolforger commented:
This is done; I'll squash commits after final review.

jcrist commented on sympy/assumptions/ask.py:
```diff
@@ -134,7 +135,11 @@ def ask(proposition, assumptions=True, context=global_assumptions):
+    known_facts_dict = get_known_facts_dict()
+
+    if local_facts \
+    and satisfiable(And(local_facts, known_facts_cnf)) is False:
```
I'd prefer if this was line broken as:
Why did you line break this? With out the linebreak, it's still only 79 characters long...

If you must do a line break, please indent two levels to indicate that it's for formatting (as per the pep); having it be on the same level as the if just looks weird. I'd prefer to break it in the parens, and ditch the backslash entirely, but that's just me.

toolforger replied:
Ah. It used to be too long, then I shortened the second condition.
Unwrapped.

jcrist commented on sympy/assumptions/ask.py:
```diff
@@ -154,7 +159,8 @@ def ask(proposition, assumptions=True, context=global_assumptions):
             return False
-    elif local_facts.func is And and all(k in known_facts_dict for k in local_facts.args):
+    elif local_facts.func is And \
+    and all(k in known_facts_dict for k in local_facts.args):
```
same here. I'd break this as:

```python
    elif local_facts.func is And and all(k in known_facts_dict
            for k in local_facts.args):
        ...
```
or something like that.

toolforger replied:
I prefer to break at the outermost level. Tearing some inner pair of parenthese apart just to avoid the backslash is "letters over structure", which I avoid because it makes understanding the code structure harder.

So I'm going for either

```
    elif (local_facts.func is And and
            all(k in known_facts_dict for k in local_facts.args)):
```

or

```
    elif local_facts.func is And and \
            all(k in known_facts_dict for k in local_facts.args):
```

Here a line break is required because the line is over 80 characters.

jcrist commented:
Besides my two formatting nitpicks, this looks to go. It appears to have merge conflicts though - will merge once those are dealt with. Thanks!

toolforger committed:
Prevent import cycles via assumptions.ask

ask.py would collect known facts during import, calling back into SymPy.
Replaced the statements with cached functions that would do it, so the
collection passes will happen only after SymPy is fully loaded.

The exec-based consistency test would not work for a function-based
facts collection; for some reason, it would not grok the import of
Or (and most likely other symbols).
I suspect that the global namespace is lost somewhere along the way
from exec String to cache to function, but I could not find that out;
instead, I resorted to a textual compare of ask_generated.py with what
compute_known_facts() returns.
If the consistency test is just to prevent an outdated ask_generated.py,
then that works, too. If the original consistency test prevented other
mishaps not tested for elsewhere, then that needs fixing.

toolforger commented:
Okay, tests passed (except for the common Theano timeout).
I rebased & squashed, that may have introduced errors so another test run needs to pass.
